polygeist/lib/Dialect/Polygeist/Transforms/KernelDisjointSpecialization.cpp

```diff
@@ -0,0 +1,289 @@
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument. Currently, all arguments
+/// with valid memref type determined by isValidMemRefType is a candidate
+/// argument, i.e., 'memref<?x!sycl.accessor>`.
+static bool isCandidateArg(Value arg) {
+  if (isValidMemRefType(arg.getType())) {
+    assert(isa<AccessorPtrType>(arg) &&
+           "Expecting valid memref type to be AccessorPtrType");
+    return true;
+  }
+  return false;
+}
+
+/// Populates \p candArgs with candidate arguments from \p call.
+static void
+collectCandidateArguments(CallOpInterface call,
+                          SmallVectorImpl<AccessorPtrType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArg(arg))
+      candArgs.push_back(cast<AccessorPtrType>(arg));
+}
+
+/// Updates \p newFnName to a unique function name if it is already defined.
+static void getUniqueFnName(std::string &newFnName, Operation *symbolTable) {
+  assert(symbolTable && symbolTable->hasTrait<OpTrait::SymbolTable>() &&
+         "Expecting symbol table");
+  auto alreadyDefined = [&symbolTable](std::string fnName) {
+    return llvm::any_of(symbolTable->getRegion(0), [&](auto &block) {
+      return llvm::any_of(block, [&](auto &op) {
+        auto nameAttr = op.template getAttrOfType<StringAttr>(
+            SymbolTable::getSymbolAttrName());
+        return nameAttr && nameAttr.getValue() == fnName;
+      });
+    });
+  };
+  unsigned counter = 0;
+  while (alreadyDefined(newFnName))
+    newFnName += ("." + std::to_string(counter));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  std::string newFnName = (func.getName() + ".specialized").str();
+  getUniqueFnName(newFnName, func->getParentWithTrait<OpTrait::SymbolTable>());
+  OpBuilder builder(func);
+  FunctionOpInterface clonedFunc = func.clone();
+  clonedFunc.setName(newFnName);
+  builder.insert(clonedFunc);
+  polygeist::privatize(clonedFunc);
+  return clonedFunc;
+}
+
+/// Adds attribute 'sycl.inner.disjoint' to arguments of \p func that satisfy \p
+/// predicate.
+template <typename Pred,
+          typename = std::enable_if_t<std::is_invocable_r_v<bool, Pred, Value>>>
+static void setInnerDisjointAttribute(FunctionOpInterface func,
+                                      Pred predicate) {
+  constexpr StringLiteral innerDisjointAttrName = "sycl.inner.disjoint";
```

Comment:
This will change to get the correct attribute name from the SYCL dialect (in a future PR).